modules/videoio/src/cap_ffmpeg_hw.hpp

```diff
@@ -473,9 +476,16 @@ void hw_init_opencl(AVBufferRef* ctx) {
     if (!hw_device_ctx)
         return;
 #ifdef HAVE_VA_INTEL
-    VADisplay va_display = hw_get_va_display(hw_device_ctx);
-    if (va_display) {
-        va_intel::ocl::initializeContextFromVA(va_display);
+    ocl::OpenCLExecutionContext& ocl_context = ocl::OpenCLExecutionContext::getCurrent();
+    cv::va_intel::VAAPIInterop* interop = ocl_context.getContext().getUserContext<cv::va_intel::VAAPIInterop>().get();
+    //only initialize the context automatically if it isn't already
+    if(!interop) {
+        VADisplay va_display = hw_get_va_display(hw_device_ctx);
+        if (va_display) {
+            va_intel::ocl::initializeContextFromVA(va_display);
+        }
+    } else {
+        CV_LOG_DEBUG(NULL, "OpenCL/VA_INTEL: CL/VA interop already initialized. ")
```

Comment:
If there is `HW_ACCELERATION_USE_OPENCL=1` then we should initialize context anyway (as this is explicitly requested by user through dedicated property).

---

Is there a way to check if OpenCL context could serve for compatible (but different) `VADisplay`s?

Reply:
I try to find one but initializing anyway would lead to the situation we had with different VADisplays and no VA-capability

Reply:
Also, when the user sets ```HW_ACCELERATION_USE_OPENCL=1``` on both VideoCapture and VideoWriter in the same program tha would lead again to the situation with two different VADisplays.

Reply:
So this wouldn't benefit from VA:
```c++
   cv::VideoCapture capture("in.mkv", cv::CAP_FFMPEG, {
            cv::CAP_PROP_HW_DEVICE, VA_HW_DEVICE_INDEX,
            cv::CAP_PROP_HW_ACCELERATION, cv::VIDEO_ACCELERATION_VAAPI,
            cv::CAP_PROP_HW_ACCELERATION_USE_OPENCL, 1
    });

    cv::VideoWriter writer("out.mkv", cv::CAP_FFMPEG, cv::VideoWriter::fourcc('V', 'P', '9', '0'), fps, cv::Size(width, height), {
            cv::VIDEOWRITER_PROP_HW_ACCELERATION, cv::VIDEO_ACCELERATION_VAAPI,
            cv::VIDEOWRITER_PROP_HW_ACCELERATION_USE_OPENCL, 1
    });

    cv::Umat frame;
    capture >> frame;
    write << frame;
```

full example: https://github.com/kallaballa/OpenCV-Issues/blob/main/src/vaapi-decode-encode/vaapi-decode-encode.cpp